asim: extend datadriven test to support recovery

kvoli opened the pull request:
Previously, only rebalancing was supported in the data driven simulation
test. This commit extends the syntax to support recovery scenarios. As
part of the extension, the state generator is split into range
generation and cluster generation.

Examples are added for each command, along with common testing scenarios
such as decommissioning, IO overload, disk fullness and adding a node
(with a store).

The newly supported commands are listed below:

```
- "load_cluster": config=<name>
  Load a defined cluster configuration to be the generated cluster in the
  simulation. The available confiurations are: single_region: 15 nodes in
  region=US, 5 in each zone US_1/US_2/US_3. single_region_multi_store: 3
  nodes, 5 stores per node with the same zone/region configuration as
  above. multi_region: 36 nodes, 12 in each region and 4 in each zone,
  regions having 3 zones. complex: 28 nodes, 3 regions with a skewed
  number of nodes per region.

- "gen_ranges" [ranges=<int>] [placement_skew=<bool>] [repl_factor=<int>]
  [keyspace=<int>] [range_bytes=<int>]
  Initialize the range generator parameters. On the next call to eval, the
  range generator is called to assign an ranges and their replica
  placement. The default values are ranges=1 repl_factor=3
  placement_skew=false keyspace=10000.

- set_liveness node=<int> [delay=<duration>]
  status=(dead|decommissioning|draining|unavailable)
  Set the liveness status of the node with ID NodeID. This applies at the
  start of the simulation or with some delay after the simulation starts,
  if specified.

- add_node: [stores=<int>] [locality=<string>] [delay=<duration>]
  Add a node to the cluster after initial generation with some delay,
  locality and number of stores on the node. The default values are
  stores=0 locality=none delay=0.

- set_span_config [delay=<duration>]
  [startKey, endKey): <span_config> Provide a new line separated list
  of spans and span configurations e.g.
  [0,100): num_replicas=5 num_voters=3 constraints={'+region=US_East'}
  [100, 500): num_replicas=3
  ...
  This will update the span config for the span [0,100) to specify 3
  voting replicas and 2 non-voting replicas, with a constraint that all
  replicas are in the region US_East.

- assertion extended to support two new assertion types:
  For type=stat assertions, if the stat (e.g. stat=replicas) value of the
  last ticks (e.g. ticks=5) duration is not exactly equal to threshold,
  the assertion fails. This applies for a specified store which must be
  provided with store=storeID.

  For type=conformance assertions, you may assert on the number of
  replicas that you expect to be under-replicated(under),
  over-replicated(over), unavailable(unavailable) and violating
  constraints(violating) at the end of the evaluation.

- "topology" [sample=<int>]
  Print the cluster locality topology of the sample given (default=last).
  e.g. for the load_cluster config=single_region
  US
  ..US_1
  ....└── [1 2 3 4 5]
  ..US_2
  ....└── [6 7 8 9 10]
  ..US_3
  ....└── [11 12 13 14 15]
```

Informs: #90137
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100979)
<!-- Reviewable:end -->

kvoli committed:
asim: add capacity override

This commit enables setting a capacity override on the state object for
a given store. The intention with this change is to enable open-loop
(without feedback changing the capacity value) simulation of certain
store capacity fields, such as IO overload.

Release note: None

kvoli committed:
asim: add config change listener

This commit adds a listener for cluster configuration changes such as
adding a new store. The configuration change listener
(`ConfigChangeListener`) will be called each time a new store is added
to the simulation state. While not immediately useful, the intention is
to use this in conjunction with events to enable adding stores after the
initial cluster state has been initialized.

Release note: None

kvoli committed:
asim: add delayed events

Previously, there was no method to alter the simulation state once the
simulation had begun executing. This commit adds delayed events, which
execute a static duration after the simulation starts, on the simulation
state. These events may be used for adding stores or altering the
liveness status of nodes for example.

Release note: None

kvoli committed:
asim: add state conformance report

This commit adds the ability to generate a conformance  report on the
simulated cluster state. The conformance report is the same that is used
for an actual cluster, with the necessary dependencies mocked for
simulation.

The conformance report allows discerning unavailable, under-replicated,
over-replicated and constraint violating ranges. This functionality may
be used to check for potential bugs in the allocation code.

Release note: None

kvoli committed:
asim: add disk-fraction-used metric

This commit adds simulation metric tracking for each store's fraction of
disk being used.

Release note: None

kvoli committed:
asim: extend datadriven test to support recovery

Previously, only rebalancing was supported in the data driven simulation
test. This commit extends the syntax to support recovery scenarios. As
part of the extension, the state generator is split into range
generation and cluster generation.

Examples are added for each command, along with common testing scenarios
such as decommissioning, IO overload, disk fullness and adding a node
(with a store).

The newly supported commands are listed below:

```
- "load_cluster": config=<name>
  Load a defined cluster configuration to be the generated cluster in the
  simulation. The available confiurations are: single_region: 15 nodes in
  region=US, 5 in each zone US_1/US_2/US_3. single_region_multi_store: 3
  nodes, 5 stores per node with the same zone/region configuration as
  above. multi_region: 36 nodes, 12 in each region and 4 in each zone,
  regions having 3 zones. complex: 28 nodes, 3 regions with a skewed
  number of nodes per region.

- "gen_ranges" [ranges=<int>] [placement_skew=<bool>] [repl_factor=<int>]
  [keyspace=<int>] [range_bytes=<int>]
  Initialize the range generator parameters. On the next call to eval, the
  range generator is called to assign an ranges and their replica
  placement. The default values are ranges=1 repl_factor=3
  placement_skew=false keyspace=10000.

- set_liveness node=<int> [delay=<duration>]
  status=(dead|decommissioning|draining|unavailable)
  Set the liveness status of the node with ID NodeID. This applies at the
  start of the simulation or with some delay after the simulation starts,
  if specified.

- add_node: [stores=<int>] [locality=<string>] [delay=<duration>]
  Add a node to the cluster after initial generation with some delay,
  locality and number of stores on the node. The default values are
  stores=0 locality=none delay=0.

- set_span_config [delay=<duration>]
  [startKey, endKey): <span_config> Provide a new line separated list
  of spans and span configurations e.g.
  [0,100): num_replicas=5 num_voters=3 constraints={'+region=US_East'}
  [100, 500): num_replicas=3
  ...
  This will update the span config for the span [0,100) to specify 3
  voting replicas and 2 non-voting replicas, with a constraint that all
  replicas are in the region US_East.

- assertion extended to support two new assertion types:
  For type=stat assertions, if the stat (e.g. stat=replicas) value of the
  last ticks (e.g. ticks=5) duration is not exactly equal to threshold,
  the assertion fails. This applies for a specified store which must be
  provided with store=storeID.

  For type=conformance assertions, you may assert on the number of
  replicas that you expect to be under-replicated(under),
  over-replicated(over), unavailable(unavailable) and violating
  constraints(violating) at the end of the evaluation.

- "topology" [sample=<int>]
  Print the cluster locality topology of the sample given (default=last).
  e.g. for the load_cluster config=single_region
  US
  ..US_1
  ....└── [1 2 3 4 5]
  ..US_2
  ....└── [6 7 8 9 10]
  ..US_3
  ....└── [11 12 13 14 15]
```

Release note: None

kvoli commented:
Rebased on master.

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 2 of 30 files at r27, 31 of 57 files at r34, 6 of 27 files at r35, 4 of 26 files at r36, 15 of 24 files at r37, 2 of 23 files at r38, 1 of 22 files at r39, 15 of 33 files at r40, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-:-NY-b1bP-CKRtDA4WkDs:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kvoli)

___
*[`-- commits` line 6 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NXzi5c5B6x0_U5JqiqL):*
nit: certain
___
*[`pkg/kv/kvserver/asim/gossip/gossip.go` line 184 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NXznRptDv08YWIowxFq:-NXznRptDv08YWIowxFr:blt4bt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/41a453a73ac6b505310efe33896631db77826b0e/pkg/kv/kvserver/asim/gossip/gossip.go#L184)):*
> ```Go
> 	if sg, ok := g.storeGossip[storeID]; ok {
> 		if !sg.addingStore {
> 			sg.local.UpdateCachedCapacity(capacity)
> ```

why is this gated on `!sg.addingStore`. How will the new capacity be communicated given that we choose to ignore it here?
___
*[`pkg/kv/kvserver/asim/state/impl.go` line 1177 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-SpDrAzLcxm7Dr5ip:-NY-SpDrAzLcxm7Dr5iq:b-5d89e3) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a52a42b756ac314b290e885b11116efde3c0b2/pkg/kv/kvserver/asim/state/impl.go#L1177)):*
> ```Go
> 			entry.IsLive = false
> 		case livenesspb.NodeLivenessStatus_UNAVAILABLE:
> 			entry.Liveness.Expiration.WallTime = s.clock.Now().Add(-time.Second).UnixNano()
> ```

Is this choice of 1s in the past important? Could it have been set to 0, like the preceding DEAD case?
___
*[`pkg/kv/kvserver/asim/state/impl.go` line 1184 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-T4E3344Jdyg6wm1L:-NY-T4E3344Jdyg6wm1M:bk7wanx) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a52a42b756ac314b290e885b11116efde3c0b2/pkg/kv/kvserver/asim/state/impl.go#L1184)):*
> ```Go
> 		case livenesspb.NodeLivenessStatus_DECOMMISSIONED:
> 			entry.Membership = livenesspb.MembershipStatus_DECOMMISSIONED
> 			entry.Liveness.Expiration.WallTime = int64(0)
> ```

Similar question on the 0 value here. Could it be anything in the past?

Code comments for these constants would be helpful.
___
*[`pkg/kv/kvserver/asim/state/impl.go` line 1218 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-TNRfC6zEVLxdarL_:-NY-TNRfC6zEVLxdarLa:b-h60hu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a52a42b756ac314b290e885b11116efde3c0b2/pkg/kv/kvserver/asim/state/impl.go#L1218)):*
> ```Go
> 
> // GetSpanConfigForKey is added for the spanconfig.StoreReader interface, required for
> // SpanConfigConformanceReport.
> ```

So some methods needed for `SpanConfigConformanceReport` are implemented and some call `panic("not implemented")`. This is confusing.
___
*[`pkg/spanconfig/spanconfigtestutils/utils.go` line 653 at r39](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-VqOY1sKacAWvuGLH:-NY-VqOY1sKacAWvuGLI:bdy62pa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ece4886564e41c3ee28f2c78f62db90941e0adc/pkg/spanconfig/spanconfigtestutils/utils.go#L653)):*
> ```Go
> 	}
> 	if !reflect.DeepEqual(conf.LeasePreferences, defaultConf.LeasePreferences) {
> 		diffs = append(diffs, fmt.Sprintf("lease_preferences=%v", conf.LeasePreferences))
> ```

no test output changed because of this fix? Should a test be added?
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_add_node` line 19 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-YETs8T2YqFInxRt-:-NY-YETs8T2YqFInxRt0:b-yp4bat) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_add_node#L19)):*
> ```
> ----
> 
> # Assert that the replica counts balance within 5% of eachother among stores.
> ```

nit: each other
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_add_node` line 33 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-YUA3Eg8lx_YTAqvv:-NY-YUA3Eg8lx_YTAqvw:b4z9hrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_add_node#L33)):*
> ```
> OK
> 
> plot stat=replicas sample=1
> ```

is this plotting the replica count for one of the stores, to show that the store will get to 300 replicas because each of the 300 ranges need 3 replicas?
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 3 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-ZNtpBShcALekkR3R:-NY-ZNtpBShcALekkR3S:b-u9uf4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L3)):*
> ```
> # This example sets n7 to dead initially and n5 to decommissioning after 2
> # minutes. The output of replicas per store is then plotted. There are no
> # assertions made.
> ```

There are assertions being made below.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 18 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-_B-M7CB05g5F9JaD:-NY-_B-M7CB05g5F9JaE:b-vmwer2) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L18)):*
> ```
> ----
> 
> # n6 becomes decommissioning after 2 minutes and remains decommissioning
> ```

3 minutes?
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 24 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-Zn7fDKJPDob8Ub0N:-NY-Zn7fDKJPDob8Ub0O:bk7cone) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L24)):*
> ```
> 
> # The number of replicas on the dead store should be 0, assert this.
> assertion type=stat stat=replicas ticks=6 threshold=0 store=7
> ```

is each tick 500ms (`defaultTickInterval`)? In 6 ticks there will be no replicas left, so 3000 ranges will be moved? I expected it to be slower.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 36 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-_GeiB2Doptx27TiT:-NY-_GeiB2Doptx27TiU:b-ud5swz) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L36)):*
> ```
> 
> # We expect one node(store) (n7) to immediately start losing replicas, whilst
> # other stores gain replicas evenly. After 5 minutes, we expect another
> ```

why after 5 minutes, given it has transitioned to decommissioning after 3 min? What is happening in the interim 2 min?
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_load_cluster` line 12 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-_bc46hBta_Umrq9n:-NY-_bc46hBta_Umrq9o:b92j8r5) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_load_cluster#L12)):*
> ```
> 
> # Set the span config so that there are only voters, with 3 voters in US_East
> # and 1 voter in US_West and EU.
> ```

nit: 1 voter each in ...
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_multi_store` line 4 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-a0woARfSvwU_KmRh:-NY-a0woARfSvwU_KmRi:b-uik4kz) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_multi_store#L4)):*
> ```
> # test. The number of leases per store should be equal to 1. We assert on this
> # with a balance threshold of 1 (i.e. identical number of leases) and a steady
> # state threshold of 0 (i.e. doesn't change).
> ```

Can you elaborate on what a "steady state threshold" means? Each store has 1 lease, so I don't know what a 0 in this case means.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_multi_store` line 5 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-aT2R3nE-ukBmjSvy:-NY-aT2R3nE-ukBmjSvz:br4jiah) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_multi_store#L5)):*
> ```
> # with a balance threshold of 1 (i.e. identical number of leases) and a steady
> # state threshold of 0 (i.e. doesn't change).
> gen_cluster stores=7 stores_per_node=2
> ```

`stores=7 stores_per_node=2` means 7 stores to me. If this is configuring 14 stores shouldn't the directive be `nodes=7 stores_per_node=2`?
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_multi_store` line 28 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-aI2NCok0zISiF70R:-NY-aI2NCok0zISiF70S:bvpx08n) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_multi_store#L28)):*
> ```
> ----
> 
>  14.00 ┼╮
> ```

Are all the leases with 1 store initially? Why? 


<!-- Sent from Reviewable.io -->

sumeerbhola commented on pkg/kv/kvserver/asim/tests/testdata/example_fulldisk:
```diff
@@ -0,0 +1,63 @@
+
+plot stat=replicas
+----
+----
```
Could you add a comment on what this graph is showing.

sumeerbhola commented on pkg/kv/kvserver/asim/tests/testdata/example_io_overload:
```diff
@@ -0,0 +1,39 @@
+----
+OK
+
+plot stat=replicas
```
what is this test trying to do?

kvoli reviewed (commented):
TYFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-:-NY_JdIfDlBiuLHqFv7L:bzftgo7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @kvoli and @sumeerbhola)

___
*[`-- commits` line 6 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NXzi5c5B6x0_U5JqiqL):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: certain
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/asim/gossip/gossip.go` line 184 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NXznRptDv08YWIowxFq:-NY_2OGm3dMo2hroBHaL:b-z184u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/41a453a73ac6b505310efe33896631db77826b0e/pkg/kv/kvserver/asim/gossip/gossip.go#L184)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is this gated on `!sg.addingStore`. How will the new capacity be communicated given that we choose to ignore it here?
</blockquote></details>

This is to avoid recursive calls. We grab the descriptor, but this will trigger a capacity change gossip. So until the new store gossip is written to the map, we use a dummy one with `addingStore=true`. This is a tad hacky.
___
*[`pkg/kv/kvserver/asim/state/impl.go` line 1177 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-SpDrAzLcxm7Dr5ip:-NY_8J_615Gt9JUUkvKf:bpj3w5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a52a42b756ac314b290e885b11116efde3c0b2/pkg/kv/kvserver/asim/state/impl.go#L1177)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Is this choice of 1s in the past important? Could it have been set to 0, like the preceding DEAD case?
</blockquote></details>

The expiration timestamp needs to be within now – `server.time_until_store_dead` which is default 5 minutes, otherwise it would be dead.
___
*[`pkg/kv/kvserver/asim/state/impl.go` line 1184 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-T4E3344Jdyg6wm1L:-NY_AIXXB6kO1-ti_pLl:bbd0bvh) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a52a42b756ac314b290e885b11116efde3c0b2/pkg/kv/kvserver/asim/state/impl.go#L1184)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Similar question on the 0 value here. Could it be anything in the past?

Code comments for these constants would be helpful.
</blockquote></details>

Added comments to constants.
___
*[`pkg/kv/kvserver/asim/state/impl.go` line 1218 at r37](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-TNRfC6zEVLxdarL_:-NY_AZlD5zOnNz8LxAEG:bierwkl) ([raw file](https://github.com/cockroachdb/cockroach/blob/97a52a42b756ac314b290e885b11116efde3c0b2/pkg/kv/kvserver/asim/state/impl.go#L1218)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

So some methods needed for `SpanConfigConformanceReport` are implemented and some call `panic("not implemented")`. This is confusing.
</blockquote></details>

The ones which are actually used for SpanConfigConformanceReport are filled in. The others add a panic. I've added comments to these mentioning they're not used for SpanConfigConformanceReport but need to be there to satisfy the (wide) iface.
___
*[`pkg/spanconfig/spanconfigtestutils/utils.go` line 653 at r39](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-VqOY1sKacAWvuGLH:-NY_CJOn0xJNOJTHXeu2:b-d6qngl) ([raw file](https://github.com/cockroachdb/cockroach/blob/1ece4886564e41c3ee28f2c78f62db90941e0adc/pkg/spanconfig/spanconfigtestutils/utils.go#L653)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

no test output changed because of this fix? Should a test be added?
</blockquote></details>

It isn't really being used by the span config conformance reports. The lease preference is just a preference, so not included. I'm hijacking the methods here, so I thought I'd add it in – but now realizing we don't even use. I'll drop this commit for now.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_add_node` line 19 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-YETs8T2YqFInxRt-:-NY_CrZ9E5ftU8UjEQM-:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_add_node#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: each other
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_add_node` line 33 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-YUA3Eg8lx_YTAqvv:-NY_CttC-Ghk5AA4qb-b:b862qq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_add_node#L33)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is this plotting the replica count for one of the stores, to show that the store will get to 300 replicas because each of the 300 ranges need 3 replicas?
</blockquote></details>

Yup, added a comment mentioning this.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_fulldisk` line 19 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#gh-1231518278:-NY_EmCnEEWJOZ2itdFZ:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_fulldisk#L19)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Could you add a comment on what this graph is showing.
</blockquote></details>

Added.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_io_overload` line 17 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#gh-1231519249:-NY_F337BXdw2n5ilPhM:b-p86sdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_io_overload#L17)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what is this test trying to do?
</blockquote></details>

Show that a IO overloaded store won't be rebalanced towards. Added a comment.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 3 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-ZNtpBShcALekkR3R:-NY_FAq9Bhxa5pymyB3L:btsnpcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L3)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

There are assertions being made below.
</blockquote></details>

Indeed, removed this comment.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 18 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-_B-M7CB05g5F9JaD:-NY_FjNN-f7vqoyuUWeJ:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L18)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

3 minutes?
</blockquote></details>

Fixed.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 24 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-Zn7fDKJPDob8Ub0N:-NY_Fpul7tscnUg5UgpR:bifp1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L24)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

is each tick 500ms (`defaultTickInterval`)? In 6 ticks there will be no replicas left, so 3000 ranges will be moved? I expected it to be slower.
</blockquote></details>

Each tick is the metrics interval. So 10s by default. The assertion is made over the last 60s of logical time.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_liveness` line 36 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-_GeiB2Doptx27TiT:-NY_GEIO2DlkJVER7znO:bbgi5v) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_liveness#L36)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why after 5 minutes, given it has transitioned to decommissioning after 3 min? What is happening in the interim 2 min?
</blockquote></details>

Yeah sorry, this should be 3 minutes as well. Idk where my times got all messed up from. Fixed.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_load_cluster` line 12 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-_bc46hBta_Umrq9n:-NY_GOlYBd8h0aLrHu3S:b-bw8e14) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_load_cluster#L12)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

nit: 1 voter each in ...
</blockquote></details>

Updated.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_multi_store` line 4 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-a0woARfSvwU_KmRh:-NY_GSLU4Q8uJinAqaRj:b-34i49s) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_multi_store#L4)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Can you elaborate on what a "steady state threshold" means? Each store has 1 lease, so I don't know what a 0 in this case means.
</blockquote></details>

It means that each store's lease count doesn't change by more than 0 in the last 6 ticks (60s). The balance assertion is max/mean, so max = mean to pass this test as well.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_multi_store` line 5 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-aT2R3nE-ukBmjSvy:-NY_HJxc4nI0CIgSWtel:bd1cqq9) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_multi_store#L5)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

`stores=7 stores_per_node=2` means 7 stores to me. If this is configuring 14 stores shouldn't the directive be `nodes=7 stores_per_node=2`?
</blockquote></details>

That is a good point. I renamed to `nodes` and `stores_per_node` in this commit.
___
*[`pkg/kv/kvserver/asim/tests/testdata/example_multi_store` line 28 at r40](https://reviewable.io/reviews/cockroachdb/cockroach/100979#-NY-aI2NCok0zISiF70R:-NY_HNsLB4g_oPppPQ4V:b-o0fa1q) ([raw file](https://github.com/cockroachdb/cockroach/blob/928f0ed46be2f543ded976661916f7bac24c8cb5/pkg/kv/kvserver/asim/tests/testdata/example_multi_store#L28)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Are all the leases with 1 store initially? Why? 
</blockquote></details>

Generated that way.


<!-- Sent from Reviewable.io -->

kvoli commented:
bors r=sumeerbhola

kvoli commented:
bors r+

kvoli commented:
bors r+
